### ISSUE TITLE
How to do multi-task learning in AllenNLP?

### ISSUE THREAD

**Issue by tingkai-zhang** [allenai/allennlp#2618]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.7
 - AllenNLP version: v0.8.2

**Question**
What does a typical multi-task learning configuration file (using AllenNLP later than v0.8.3) look like?

Because different sub-tasks have different suitable learning rate range.

How can I config different learning rates for different sub-tasks?

Thank you!

**Comment by matt-gardner** [comment]
One somewhat hacky way of doing this is to just multiply your loss by a constant in the model depending on which task you're doing at the time - this should change the magnitude of the gradients in the same way that a different learning rate would.  I don't think we currently have a way of setting different learning rates otherwise, though others might know this better than me.

**Comment by brendan-ai2** [comment]
@joelgrus, FYI. This may be relevant to your multi-task work from earlier in the year.

**Comment by kl2806** [comment]
Any thoughts on this, @joelgrus ?

**Comment by joelgrus** [comment]
right now there's not an official multitask trainer in AllenNLP, just the toy example one:

https://github.com/allenai/allennlp/blob/master/allennlp/tests/training/multi_task_trainer_test.py

you could accomplish this with two small changes:

1. instead of passing the trainer an `Optimizer`, pass it a `Dict[str, Optimizer]`  where the keys are the names of the datasets

https://github.com/allenai/allennlp/blob/master/allennlp/tests/training/multi_task_trainer_test.py

2. then in the `train` method, choose which optimizer to use based on the batch type:

https://github.com/allenai/allennlp/blob/master/allennlp/tests/training/multi_task_trainer_test.py#L225

I think we're still figuring out what a *general* multitask trainer would look like, which is one reason why we haven't implemented it. (also the proposed "callbacks" work would make it a lot easier, otherwise we'd have to duplicate all the logging + checkpointing + etc code)

**Comment by sedflix** [comment]
this might be helpful: https://github.com/huggingface/hmtl
